osu.Game/Extensions/DrawableExtensions.cs

```diff
@@ -79,5 +80,42 @@ public static void ApplySkinnableInfo(this Drawable component, SkinnableInfo inf
                     container.Add(child.CreateInstance());
             }
         }
+
+        /// <summary>
+        /// Keeps the drawable upright and prevents it from being scaled or flipped with its Parent.
+        /// </summary>
+        /// <param name="drawable">The drawable.</param>
+        public static void KeepUprightAndUnscaled(this Drawable drawable)
+        {
+            var parentMatrix = drawable.Parent.DrawInfo.Matrix;
+            float angle = MathF.Atan(parentMatrix.M12 / parentMatrix.M11);
+            angle = MathHelper.RadiansToDegrees(angle);
+
+            parentMatrix.Transpose();
+            parentMatrix.M13 = 0.0f;
+            parentMatrix.M23 = 0.0f;
+
+            if ((Math.Abs(Math.Abs(angle) - 90.0)) < 2.0f)
+            {
+                Matrix3 m = Matrix3.CreateRotationZ(MathHelper.DegreesToRadians(40.0f));
+                m.Transpose();
+                parentMatrix *= m;
+                drawable.Rotation = 40.0f;
+            }
+            else
+                drawable.Rotation = 0.0f;
+
+            Matrix3 C = parentMatrix.Inverted();
+
+            float alpha, beta, sx, sy;
+            sy = C.M22;
+            alpha = C.M12 / C.M22;
+
+            beta = (C.M21 == 0.0f) ? 0.0f : 1 / ((C.M11 / C.M21) - alpha);
+            sx = (beta == 0.0f) ? C.M11 : C.M21 / beta;
```

Comment:
This needs to be documented better. I assume you're doing something like decomposing affine transforms here but it's unclear what is going on exactly.

Reply:
in the previous implementation, the Rotation and Scale of the drawable that we want to keep upright and unscaled is set to the inverse of the rotation and scale of its parent, which doesn't work because R(theta)*S(s)*R(-theta)*S(1/s) is not equal to the identity matrix, this implementation uses shear and scale instead of rotation and scale.

in the lines you mentioned sx and sy are the scale factors, alpha and beta are shear factors and the values being asingned are the solution of the system of equations that we need to solve to find the right factors.